api/python/t4/formats.py

```diff
@@ -0,0 +1,1007 @@
+    opts = ('doublequote', 'encoding', 'escapechar', 'fieldsep', 'header_names', 'index_names',
+            'index_names_are_keys', 'linesep', 'na_values', 'quotechar', 'quoting', 'skip_spaces', 'use_header',
+            'use_index')
+    # defaults shouldn't be added to metadata, just used directly.
+    defaults = {
+        'encoding': 'utf-8',
+        'index_names_are_keys': False,
+        'na_values': ['', '#N/A', '#N/A N/A', '#NA',
+            '-1.#IND', '-1.#QNAN', '-NaN', '-nan', '1.#IND', '1.#QNAN',
+            'N/A', 'NA', 'NULL', 'NaN', 'n/a', 'nan', 'null'],
+        'use_header': True,
+        'use_index': False,
+    }
+
+    def handles_type(self, typ):
+        # don't load pandas unless we actually have to use it..
+        if 'pandas' not in sys.modules:
+            return False
+        import pandas as pd
+
+        self.handled_types.add(pd.DataFrame)
+
+        return super().handles_type(typ)
+
+    def _quoting_opt_to_python(self, value):
+        if isinstance(value, int):
+            return value
+        elif isinstance(value, str):
+            value = value.strip().lower()
+            map = {
+                'all': csv.QUOTE_ALL,
+                'minimal': csv.QUOTE_MINIMAL,
+                'none': csv.QUOTE_NONE,
+                'nonnumeric': csv.QUOTE_NONNUMERIC
+            }
+            return map.get(value, NOT_SET)
+        warnings.warn("Unrecognized value for 'quoting' option: {!r} (ignored)".format(value))
+        return NOT_SET
+
+    def get_ser_kwargs(self, opts):
+        opts = copy.deepcopy(opts)
+        result_kwargs = {}
+
+        # interdependent opts, can't be processed individually.
+        use_header = opts.pop('use_header')    # must exist, at least as a default
+        header_names = opts.pop('header_names', None)
+        if use_header:
+            result_kwargs['header'] = header_names if header_names else True
+        else:
+            result_kwargs['header'] = False
+
+        # No kwarg correlate for serialization
+        opts.pop('index_names_are_keys', None)
+
+        name_map = {
+            'fieldsep': 'sep',
+            'linesep': 'line_terminator',
+            'use_index': 'index',
+            'index_names': 'index_label',
+        }
+        for name, value in opts.items():
+            if name in name_map:
+                result_kwargs[name_map[name]] = value
+            elif name == 'quoting':
+                value = self._quoting_opt_to_python(value)
+                if value is NOT_SET:
+                    continue
+                result_kwargs[name] = value
+                continue
+            elif name == 'na_values':
+                result_kwargs['na_rep'] = value[0]
+            else:
+                # exact match / pass through arg
+                result_kwargs[name] = value
+
+        return result_kwargs
+
+
+    def serialize(self, obj, meta=None, ext=None, raw_args=None, **format_opts):
```

Comment:
Remove `raw_args`; we can add it back if it comes up. It's safer to start with a minimal interface, and add things later.

Reply:
Done.